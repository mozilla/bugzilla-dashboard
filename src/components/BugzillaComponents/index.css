table thead th:nth-child(3) {
    max-width: 8rem;
}