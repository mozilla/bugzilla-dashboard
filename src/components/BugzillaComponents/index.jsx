import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import ExpandMore from '@material-ui/icons/ExpandMore';
import METRICS from '../../utils/bugzilla/metrics';

const styles = ({
  header: {
    margin: '0.5rem 0 0 0',
  },
  metric: {
    textAlign: 'center',
  },
  metricLabel: {
    textDecoration: 'underline',
  },
  icon: {
    fontSize: '1rem',
    verticalAlign: 'bottom',
  },
});

const sortByComponentName = (a, b) => a.label - b.label;

const BugzillaComponents = ({
  classes, title, bugzillaComponents, onComponentDetails,
}) => (
  bugzillaComponents.length > 0 && (
    <div>
      <h3 className={classes.header}>{title}</h3>
      <table>
        <thead>
          <tr>
            <th />
            {onComponentDetails && <th />}
            {Object.values(METRICS).map(({ label }) => (
              <th key={label} className={classes.metricLabel}>{label}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {bugzillaComponents
            .sort(sortByComponentName)
            .map(({
              label, component, product, metrics = {},
            }) => (
              <tr key={label}>
                {onComponentDetails && (
                  <td>
                    <div
                      name={label}
                      onKeyPress={onComponentDetails}
                      onClick={onComponentDetails}
                      product={product}
                      component={component}
                      role="button"
                      tabIndex="0"
                    >
                      <ExpandMore classes={{ root: classes.icon }} />
                    </div>
                  </td>
                )}
                <td>{label}</td>
                {Object.keys(METRICS).map(metric => (
                  metrics[metric] && (
                  <td key={metric} className={classes.metric}>
                    <a href={metrics[metric].link}>{metrics[metric].count}</a>
                  </td>
                  )
                ))}
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  )
);

BugzillaComponents.propTypes = {
  classes: PropTypes.shape({}).isRequired,
  title: PropTypes.string.isRequired,
  bugzillaComponents: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      product: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.string),
      ]).isRequired,
      component: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.string),
      ]).isRequired,
      metrics: PropTypes.shape({}),
    }),
  ).isRequired,
  onComponentDetails: PropTypes.func,
};

BugzillaComponents.defaultProps = {
  onComponentDetails: undefined,
};

export default withStyles(styles)(BugzillaComponents);
