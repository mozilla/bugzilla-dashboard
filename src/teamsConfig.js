const teamConfigs = {
  domCore: {
    label: 'DOM Core',
    owner: 'someone@mozilla.com',
    product: ['Core'],
    component: [
      'DOM', 'DOM: Core & HTML', 'DOM: Events', 'Drag and Drop',
      'Editor', 'Event Handling', 'HTML: Form Submission',
      'HTML: Parser', 'Keyboard: Navigation', 'Selection',
      'Serializers',
    ],
  },
  domFission: {
    label: 'DOM Fission',
    owner: 'manager@mozilla.com',
    product: ['Core', 'Toolkit'],
    component: [
      'Document Navigation', 'XBL', 'XML', 'XPConnect', 'XSLT',
    ],
  },
  workerStoreage: {
    label: 'Worker and Storage',
    owner: 'manager@mozilla.com',
    product: ['Core', 'Toolkit'],
    component: [
      'DOM: IndexedDB', 'DOM: Push Notifications', 'DOM: Quota Manager',
      'DOM: Service Workers', 'DOM: Web Payments', 'DOM: Web Storage',
      'DOM: Workers',
    ],
  },
};

export default teamConfigs;
