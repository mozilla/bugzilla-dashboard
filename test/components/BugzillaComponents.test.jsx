import React from 'react';
import renderer from 'react-test-renderer';
import BugzillaComponents from '../../src/components/BugzillaComponents';
import bugzillaComponents from '../mocks/bugzillaComponents';
import teamsConfig from '../../src/teamsConfig';

it('renders components', () => {
  const tree = renderer
    .create((
      <BugzillaComponents
        bugzillaComponents={bugzillaComponents}
        onComponentDetails={() => null}
      />
    ))
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders components bucketed as teams', () => {
  const tree = renderer
    .create((
      <BugzillaComponents
        bugzillaComponents={Object.values(teamsConfig)}
        onComponentDetails={() => null}
      />
    ))
    .toJSON();
  expect(tree).toMatchSnapshot();
});
