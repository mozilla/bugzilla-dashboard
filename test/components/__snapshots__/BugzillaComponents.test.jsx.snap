// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders components 1`] = `
<div>
  <h3
    className="BugzillaComponents-header-1"
  />
  <table>
    <thead>
      <tr>
        <th />
        <th />
        <th
          className="BugzillaComponents-metricLabel-3"
        >
          Untriaged
        </th>
        <th
          className="BugzillaComponents-metricLabel-3"
        >
          Needinfo
        </th>
        <th
          className="BugzillaComponents-metricLabel-3"
        >
          P1s
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <div
            className="DrilldownIcon-svgWrapper-4"
            name="Core::DOM: IndexedDB"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-6 DrilldownIcon-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
        </td>
        <td>
          Core::DOM: IndexedDB
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="DrilldownIcon-svgWrapper-4"
            name="Core::JavaScript Engine"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-6 DrilldownIcon-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
        </td>
        <td>
          Core::JavaScript Engine
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="DrilldownIcon-svgWrapper-4"
            name="Core::DOM: Core & HTML"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-6 DrilldownIcon-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
        </td>
        <td>
          Core::DOM: Core & HTML
        </td>
        <td
          className="BugzillaComponents-metric-2"
        >
          <a
            href="https://bugzilla.mozilla.org/buglist.cgi?component=DOM%3A%20Core%20%26%20HTML&f1=bug_severity&f2=keywords&f3=resolution&limit=0&o1=notequals&o2=notsubstring&o3=isempty&product=Core&v1=enhancement&v2=meta"
          >
            944
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="DrilldownIcon-svgWrapper-4"
            name="Toolkit::Async Tooling"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-6 DrilldownIcon-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
        </td>
        <td>
          Toolkit::Async Tooling
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders components bucketed as teams 1`] = `
<div>
  <h3
    className="BugzillaComponents-header-1"
  />
  <table>
    <thead>
      <tr>
        <th />
        <th />
        <th
          className="BugzillaComponents-metricLabel-3"
        >
          Untriaged
        </th>
        <th
          className="BugzillaComponents-metricLabel-3"
        >
          Needinfo
        </th>
        <th
          className="BugzillaComponents-metricLabel-3"
        >
          P1s
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <div
            className="DrilldownIcon-svgWrapper-4"
            name="DOM Core"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-6 DrilldownIcon-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
        </td>
        <td>
          DOM Core
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="DrilldownIcon-svgWrapper-4"
            name="DOM Fission"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-6 DrilldownIcon-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
        </td>
        <td>
          DOM Fission
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="DrilldownIcon-svgWrapper-4"
            name="Worker and Storage"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-6 DrilldownIcon-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
        </td>
        <td>
          Worker and Storage
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
